### PR TITLE
Implement critical path app with REST endpoints

### PR DESCRIPTION
**Pull request by standev**

### CODE REVIEW

**Comment by standev** [comment]
@pavel-bastecky-alice @smacharacek please review the proposed implementation

File: src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            .collect(Collectors.joining(", "));
+
+        final String taskStarts = engine.getTasks().stream()
+            .sorted(Comparator.comparingInt(TaskEvaluated::getEarlyStart))
+            .map(task -> String.format("%s=%d", task.getTaskCode(), task.getEarlyStart()))
+            .collect(Collectors.joining(", "));
+
+        assertAll(
+            () -> assertEquals(29, engine.getMaxCost()),
+            () -> assertEquals("C, D, E, B, A", taskOrder), // this doesn't look right, I'd expect the opposite order
```
Comment (standev) [inline src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java]: A bit confused here. I would expect that `C -> B -> A` dependencies relationship means that `A` needs to be done first, then `B`, then `C`.

But the algorithm seems to treat `dependencies` as the tasks that are blocked, not as the tasks that are blocking.
I'm not familiar with this area, so please advise.

Comment (pavel-bastecky-alice) [reply src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java]: Hi Stanislav, 

I need to check the code to answer correctly. But we define dependencies list as all the tasks that need to be completed before this task can start. So in the following input

```
[
    { "taskCode" : "A", "duration" : 8, "dependencies" : [ ], ...},
    { "taskCode" : "B", "duration" : 8, "dependencies" : [ "B" ], ...},
    { "taskCode" : "C", "duration" : 8, "dependencies" : [ "C" ], ...},
    { "taskCode" : "D", "duration" : 8, "dependencies" : [ "A" ], ...}
]
```

The final sequence should be something like:

```
[
    { "taskCode" : "A", "start" : 0, "finish": 8, ...},
    { "taskCode" : "B", "start" : 8, "finish": 16, ...},
    { "taskCode" : "C", "start" : 16, "finish": 24, ...},
    { "taskCode" : "D", "start" : 8, "finish": 16, ...}
]
```

So in your case I would expect the order of work to be `A, B, C` but I'm not sure if I'm seeing the final order or some intermediate part of the process at this point.

I need to check your implementation and the respective test to make sure that this is the case.

Comment (standev) [reply src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java]: Thank you for clarification! What you are seeing in the test is the final order. Based on your description, the order should have been `A -> B -> C`, and so the current result is not correct and I need to investigate it more.

- [x] investigate incorrect dependency handling

Comment (standev) [reply src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java]: The issue has been now fixed, the order of execution now matches what is expected, as per discussion.

**Pull request merged**